MCUBoot image signing will fail in a docker container without west being installed

<!--
**Notes**
Github Discussions (https://github.com/zephyrproject-rtos/zephyr/discussions)
are available to first verify that the issue is a genuine Zephyr bug and not a
consequence of Zephyr services misuse.

This issue list is only for bugs in the main Zephyr code base
(https://github.com/zephyrproject-rtos/zephyr/). If the bug is for a project
fork (such as NCS) specific feature, please open an issue in the fork project
instead.
-->

**Describe the bug**
<!--
A clear and concise description of what the bug is.

Please also mention any information which could help others to understand
the problem you're facing:
 - What target platform are you using?
 - What have you tried to diagnose or workaround this issue?
 - Is this a regression? If yes, have you been able to "git bisect" it to a
   specific commit?
 - ...
-->
For a project we are dockerizing the build for a zephyr chip using an out-of-tree application and board, but this bug was really found thru code inspection. We had west installed on the host but we didn't install it in docker build as we weren't using it and just calling cmake directly.

**To Reproduce**
<!--
Steps to reproduce the behavior:
1. mkdir build; cd build
2. cmake -DBOARD=board\_xyz
3. make
4. See error
-->
Ensure west is not installed 
```
pip3 uninstall west
```

Then we configured our build with an out-of-tree board and application this should be broken in-tree as well - note, this wants to use a sys build because mcuboot is included and just signed.

```
cmake -GNinja -B build -DAPP_DIR=myapplications/app -DBOARD=myboard -DSB_CONFIG_BOOTLOADER_MCUBOOT=y share/sysbuild
cmake --build build
```

**Expected behavior**
Build should complete

**Logs and console output**
<!--
If applicable, add console logs or other types of debug information
e.g Wireshark capture or Logic analyzer capture (upload in zip archive).
copy-and-paste text and put a code fence (\`\`\`) before and after, to help
explain the issue. (if unable to obtain text log, add a screenshot)
-->
```
   *************************
   * Running CMake for app *
   *************************

Loading Zephyr default modules (Zephyr base).
-- Application: /Volumes/Workspace/mcu-canopener/firmware/applications/app
-- CMake version: 3.28.3

… snip snip …

-- Including signing script: /Volumes/Workspace/mcu-canopener/firmware/zephyr/cmake/mcuboot.cmake
CMake Error at /Volumes/Workspace/mcu-canopener/firmware/zephyr/cmake/mcuboot.cmake:37 (message):
  Can't sign images for MCUboot: west not found.  To fix, install west and
  ensure it's on PATH.
Call Stack (most recent call first):
  /Volumes/Workspace/mcu-canopener/firmware/zephyr/cmake/mcuboot.cmake:248 (zephyr_mcuboot_tasks)
  /Volumes/Workspace/mcu-canopener/firmware/zephyr/CMakeLists.txt:1999 (include)


-- Configuring incomplete, errors occurred!
CMake Error at cmake/modules/sysbuild_extensions.cmake:514 (message):
  CMake configure failed for Zephyr project: app

  Location: /Volumes/Workspace/mcu-canopener/firmware/applications/app
Call Stack (most recent call first):
  cmake/modules/sysbuild_images.cmake:20 (ExternalZephyrProject_Cmake)
  cmake/modules/sysbuild_default.cmake:20 (include)
  /Volumes/Workspace/mcu-canopener/firmware/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:75 (include)
  /Volumes/Workspace/mcu-canopener/firmware/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:92 (include_boilerplate)
  /Volumes/Workspace/mcu-canopener/firmware/zephyr/share/sysbuild-package/cmake/SysbuildConfig.cmake:8 (include)
  template/CMakeLists.txt:10 (find_package)


-- Configuring incomplete, errors occurred!
FAILED: sysbuild-app-prefix/src/sysbuild-app-stamp/sysbuild-app-configure /Volumes/Workspace/mcu-canopener/build/firmware/sysbuild-app-prefix/src/sysbuild-app-stamp/sysbuild-app-configure 
cd /Volumes/Workspace/mcu-canopener/build/firmware/sysbuild-app && /usr/bin/cmake -E env ZEPHYR_BASE=/Volumes/Workspace/mcu-canopener/firmware/zephyr /usr/bin/cmake "-DZEPHYR_MODULES:STRING=/Volumes/Workspace/mcu-canopener/firmware/bootloader/mcuboot;/Volumes/Workspace/mcu-canopener/firmware/modules/hal/cmsis;/Volumes/Workspace/mcu-canopener/firmware/modules/hal/stm32;/Volumes/Workspace/mcu-canopener/firmware/modules/crypto/mbedtls" -DAPP_DIR:PATH=/Volumes/Workspace/mcu-canopener/firmware/applications/app -DBOARD_ROOT:PATH=/Volumes/Workspace/mcu-canopener/firmware -DBOARD:STRING=identifier_f413zh -B /Volumes/Workspace/mcu-canopener/build/firmware/sysbuild-app -S /Volumes/Workspace/mcu-canopener/firmware/zephyr/share/sysbuild && /usr/bin/cmake -E touch /Volumes/Workspace/mcu-canopener/build/firmware/sysbuild-app-prefix/src/sysbuild-app-stamp/sysbuild-app-configure
ninja: build stopped: subcommand failed.
```

**Environment (please complete the following information):**

 - OS: MacOS
 - Toolchain: Zephyr SDK in a Docker Container
 - Version: 4.0.0

**Additional context**

After code inspection - this commit stood out

```
commit c952f09a797e0b77a131b776daac25316a06051f
Author: Jamie McCrae <jamie.mccrae@nordicsemi.no>
Date:   Wed Sep 25 11:06:45 2024 +0100

    cmake: mcuboot: Use imgtool instead of west for signing
    
    Uses imgtool directly to sign images instead of calling west sign,
    this also removes the MCUBOOT_CMAKE_WEST_SIGN_PARAMS Kconfig option
    as this has no effect
    
    Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>
```

Because reading up the mcuboot.cmake there is nothing using WEST now and so when it is removed the build succeeds as expected.

```diff
diff --git a/cmake/mcuboot.cmake b/cmake/mcuboot.cmake
index 2767af807c1..8626dd1c259 100644
--- a/cmake/mcuboot.cmake
+++ b/cmake/mcuboot.cmake
@@ -32,11 +32,6 @@ function(zephyr_mcuboot_tasks)
     endif()
   endif()
 
-  if(NOT WEST)
-    # This feature requires west.
-    message(FATAL_ERROR "Can't sign images for MCUboot: west not found. To fix, install west and ensure it's on PATH.")
-  endif()
-
   foreach(file keyfile keyfile_enc)
     if(NOT "${${file}}" STREQUAL "")
       if(NOT IS_ABSOLUTE "${${file}}")
```

--- a/cmake/mcuboot.cmake
+++ b/cmake/mcuboot.cmake
@@ -31,28 +31,50 @@ function(zephyr_mcuboot_tasks)
                       "CONFIG_MCUBOOT_SIGNATURE_KEY_FILE are set, the generated build will not be "
                       "bootable by MCUboot unless it is signed manually/externally.")
       return()
+    elseif(NOT (CONFIG_BUILD_OUTPUT_BIN OR CONFIG_BUILD_OUTPUT_HEX))
+      message(FATAL_ERROR "Can't sign images for MCUboot: Neither "
+                          "CONFIG_BUILD_OUTPUT_BIN nor CONFIG_BUILD_OUTPUT_HEX "
+                          "is enabled, so there's nothing to sign.")
     endif()
-  endif()
 
-  if(NOT WEST)
-    # This feature requires west.
-    message(FATAL_ERROR "Can't sign images for MCUboot: west not found. To fix, install west and ensure it's on PATH.")
-  endif()
+    foreach(file keyfile keyfile_enc)
+      if("${${file}}" STREQUAL "")
+        continue()
+      endif()
 
-  foreach(file keyfile keyfile_enc)
-    if(NOT "${${file}}" STREQUAL "")
+      # Find the key files in the order of preference for a simple search
+      # modeled by the if checks across the various locations
+      #
+      #  1. absolute
+      #  2. application config
+      #  3. west topdir (optional when the workspace is not west managed)
+      #
       if(NOT IS_ABSOLUTE "${${file}}")
-        # Relative paths are relative to 'west topdir'.
-        set(${file} "${WEST_TOPDIR}/${${file}}")
+        if(EXISTS "${APPLICATION_CONFIG_DIR}/${${file}}")
+          set(${file} "${APPLICATION_CONFIG_DIR}/${${file}}")
+        else()
+          # Relative paths are relative to 'west topdir'.
+          #
+          # This is the only file that has a relative check to topdir likely
+          # from the historical callouts to "west" itself before using
+          # imgtool. So, this is maintained here for backward compatibility
+          #
+          if(NOT WEST OR NOT WEST_TOPDIR)
+            message(FATAL_ERROR "Can't sign images for MCUboot: west workspace undefined. "
+                                "To fix, ensure `west topdir` is a valid workspace directory.")
+          endif()
+          set(${file} "${WEST_TOPDIR}/${${file}}")
+        endif()
       endif()
 
-      if(NOT EXISTS "${${file}}" AND NOT "${CONFIG_MCUBOOT_GENERATE_UNSIGNED_IMAGE}")
-        message(FATAL_ERROR "west sign can't find file ${${file}} (Note: Relative paths are relative to the west workspace topdir \"${WEST_TOPDIR}\")")
-      elseif(NOT (CONFIG_BUILD_OUTPUT_BIN OR CONFIG_BUILD_OUTPUT_HEX))
-        message(FATAL_ERROR "Can't sign images for MCUboot: Neither CONFIG_BUILD_OUTPUT_BIN nor CONFIG_BUILD_OUTPUT_HEX is enabled, so there's nothing to sign.")
+      if(NOT EXISTS "${${file}}")
+        message(FATAL_ERROR "Can't sign images for MCUboot: can't find file ${${file}} "
+                            "(Note: Relative paths are searched through "
+                            "APPLICATION_CONFIG_DIR=\"${APPLICATION_CONFIG_DIR}\" "
+                            "and WEST_TOPDIR=\"${WEST_TOPDIR}\")")
       endif()
-    endif()
-  endforeach()
+    endforeach()
+  endif()
 
   # No imgtool, no signed binaries.
   if(NOT DEFINED IMGTOOL)

